discovery/sync_manager.go

```diff
@@ -0,0 +1,562 @@
+func (m *syncManager) chooseRandomSyncers(n int) []*gossipSyncer {
+	var syncers []*gossipSyncer
+	for _, syncer := range m.syncers {
+		if len(syncers) == n {
+			break
+		}
+
+		// A syncer is only eligible to process sync transitions if
+		// they're currently within a chansSynced state.
+		if syncer.SyncState() == chansSynced {
+			syncers = append(syncers, syncer)
+		}
+	}
+	return syncers
+}
+
+// chooseRandomSyncer returns a random non-active syncer that's eligible for a
+// sync transition.
+//
+// NOTE: It's possible for a nil value to be returned if there are no eligible
+// candidate syncers.
+//
+// NOTE: This method requires the syncersMtx read lock to be held.
+func (m *syncManager) chooseRandomSyncer() *gossipSyncer {
+	s := m.chooseRandomSyncers(1)
+	if len(s) != 1 {
+		return nil
+	}
+	return s[0]
+}
+
+// InitSyncState is called by outside sub-systems when a connection is
+// established to a new peer that understands how to perform channel range
+// queries. We'll allocate a new gossip syncer for it, and start any goroutines
+// needed to handle new queries.
+//
+// TODO(wilmer): Only mark as activeSync if this isn't a channel peer.
+func (m *syncManager) InitSyncState(peer lnpeer.Peer) {
+	m.syncersMtx.Lock()
+	defer m.syncersMtx.Unlock()
+
+	// If we already have a syncer, then we'll exit early as we don't want
+	// to override it.
+	nodeID := routing.Vertex(peer.PubKey())
+	if _, ok := m.gossipSyncer(nodeID); ok {
+		return
+	}
+
+	log.Infof("Creating new gossipSyncer for peer=%x", nodeID[:])
+
+	encoding := lnwire.EncodingSortedPlain
+	s := newGossipSyncer(gossipSyncerCfg{
+		chainHash:     m.cfg.ChainHash,
+		channelSeries: m.cfg.ChanSeries,
+		encodingType:  encoding,
+		chunkSize:     encodingTypeToChunkSize[encoding],
+		sendToPeer: func(msgs ...lnwire.Message) error {
+			return peer.SendMessage(false, msgs...)
+		},
+	}, nodeID)
+
+	// If we've yet to reach our desired number of active syncers, then
+	// we'll use this one.
+	if len(m.activeSyncers) < m.cfg.NumActiveSyncers {
+		s.syncType = uint32(activeSync)
+		m.activeSyncers[nodeID] = s
+		m.enqueueActiveSyncer(s, false)
+		return
+	}
+
+	// Otherwise, the syncer should be initialized as passive.
+	s.syncType = uint32(passiveSync)
+	s.state = uint32(waitingQueryRangeReply)
+	m.syncers[nodeID] = s
+	s.Start()
+}
+
+// PruneSyncState is called by outside sub-systems once a peer that we were
+// previously connected to has been disconnected. In this case we can stop the
+// existing gossipSyncer assigned to the peer and free up resources.
+func (m *syncManager) PruneSyncState(peer routing.Vertex) {
+	m.syncersMtx.Lock()
+	defer m.syncersMtx.Unlock()
+
+	s, ok := m.gossipSyncer(peer)
+	if !ok {
+		return
+	}
+
+	log.Infof("Removing gossipSyncer for peer=%x", peer)
+
+	// We'll start by stopping the gossipSyncer for the disconnected peer.
+	s.Stop()
+
+	// If it's a non-active syncer, then we can just exit now.
+	if s.SyncType() != activeSync {
```

Comment:
might be safe to just check
```golang
if _, ok := m.syncers[s.peerPub]; ok {
    ...
}
```
since the sync type could be mutated in another goroutine and isn't synchronized under this mutex

Reply:
Good point, fixed!